File: src/Agent.cpp
#include "Agent.h"

Agent::Agent(int observationSpace, int actionSpace, int memorySize, float epsilonStart,
             float epsilonEnd, float epsilonDecay, float gamma, float tau, int batchSize,
             double learningRate)
    : policyNetwork(observationSpace, actionSpace),
      targetNetwork(observationSpace, actionSpace),
      memory(memorySize),
      epsilonStart(epsilonStart),
      epsilonEnd(epsilonEnd),
      epsilonDecay(epsilonDecay),
      gamma(gamma),
      tau(tau),
      batchSize(batchSize),
      learningRate(learningRate),
      actionsTaken(0),
      cleanStart(false),
      numEpisodes(10)

{
}

void Agent::saveStatistics(const std::map<std::string, std::string> &data)
{
    if (statsFilePath.empty())
    {
        std::cout << "No write file specified. Statistics will not be written to a file." << std::endl;
        return;
    }
    std::ifstream file(statsFilePath);
    if (!file)
    {
        // create file
        std::ofstream newFile(statsFilePath);
        if (!newFile)
        {
            std::cerr << "Error creating file: " << statsFilePath << std::endl;
            return;
        }
        bool isFirstKey = true;
        for (const auto &entry : data)
        {
            if (!isFirstKey)
                newFile << ",";
            isFirstKey = false;
            newFile << entry.first;
        }
        newFile << std::endl;
    }

    // append file
    std::ofstream appendFile(statsFilePath, std::ios::app);
    if (!appendFile)
    {
        std::cerr << "Error opening file for appending: " << statsFilePath << std::endl;
        return;
    }
    bool isFirstValue = true;
    for (const auto &entry : data)
    {
        if (!isFirstValue)
            appendFile << ",";
        isFirstValue = false;
        appendFile << entry.second; 
    }
    appendFile << std::endl;
}

void Agent::loadModel()
{
    if (policyFilePath.empty() or targetFilePath.empty())
    {
        std::cout << "No model file specified. Will not load model." << std::endl;
        return;
    }
    if (std::filesystem::exists(policyFilePath))
    {
        torch::load(policyNetwork, policyFilePath);
    }
    if (std::filesystem::exists(targetFilePath))
    {
        torch::load(targetNetwork, targetFilePath);
    }
}

void Agent::saveModel()
{
    if (policyFilePath.empty() or targetFilePath.empty())
    {
        std::cout << "No model file specified. Will not save model." << std::endl;
        return;
    }
    torch::save(policyNetwork, policyFilePath);
    torch::save(targetNetwork, targetFilePath);
}

torch::Tensor Agent::selectAction(torch::Tensor state)
{
    // get random double
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    double sample = dis(gen);

    // calculate epsilon threshold
    double eps_threshold = epsilonEnd + (epsilonStart - epsilonEnd) * std::exp(-1.0 * actionsTaken / epsilonDecay);

    actionsTaken++;

    if (sample > eps_threshold)
    {
        torch::NoGradGuard no_grad;

        // get the output tensor from the policy net
        torch::Tensor output = policyNetwork(state);

        // find the index of the element with the largest value in each row
        torch::Tensor max_indices = std::get<1>(output.flatten().max(0));

        // reshape the tensor to have shape (1, 1)
        torch::Tensor result = max_indices.view({1, 1});
        return result;
    }
    else
        return torch::randint(env->getNumActions(), {1, 1});
}

bool Agent::act(torch::Tensor &state)
{
    torch::Tensor action = selectAction(state);
    auto [observation, reward, terminated] = env->step(action.item<int>());

    // if terminated, set tensor to zeros, else set to observation
    torch::Tensor nextState;
    if (terminated)
    {
        nextState = torch::empty({env->getNumObservations()}).unsqueeze(0);
    }
    else
    {
        nextState = observation.to(torch::kFloat32).unsqueeze(0);
    }

    // Create transition and push to memory
    Transition transition = Transition(state, nextState, action, reward);

    memory.push(transition);

    state = nextState;

    return terminated;
}

void Agent::learn(torch::optim::AdamW &optimizer)
{
    // if our memory is not large enough, don't optimize
    if (memory.size() < batchSize)
        return;
    // get a random sample of memory
    std::vector<Transition> transitions = memory.sample(batchSize);
    TransposedTransition transposedTransitions = TransposedTransition(transitions);

    // stuff
    std::vector<torch::Tensor> resultVector;
    std::vector<torch::Tensor> maskVector;

    for (const auto &tensor : transposedTransitions.nextStates)
    {
        // if tensor is empty (final state)
        if (torch::equal(tensor, torch::empty({9}).unsqueeze(0)))
        {
            resultVector.push_back(tensor);
            maskVector.push_back(torch::zeros({1}, torch::kBool));
        }
        else
        {
            resultVector.push_back(tensor);
            maskVector.push_back(torch::ones({1}, torch::kBool));
        }
    }

    torch::Tensor nonFinalMask = torch::cat(maskVector);
    torch::Tensor nonFinalNextStates = torch::cat(resultVector);

    torch::Tensor states = torch::cat(transposedTransitions.states);
    torch::Tensor actions = torch::cat(transposedTransitions.actions);
    torch::Tensor rewards = torch::cat(transposedTransitions.rewards);

    // compute Q(s_t, a)
    torch::Tensor stateActionValues = policyNetwork(states).gather(1, actions);

    // Compute V(s_{t+1}) for all next states
    torch::Tensor nextStateValues = torch::zeros(batchSize);
    {
        // Forward non-final states and get the max value for each
        torch::NoGradGuard no_grad;
        torch::Tensor nonZeroIndices = nonFinalMask.nonzero().squeeze();
        torch::Tensor targetNetworkOutputNonFinal = targetNetwork(nonFinalNextStates);
        torch::Tensor maxValuesNonFinal = std::get<0>(targetNetworkOutputNonFinal.index({nonZeroIndices}).max(1));

        nextStateValues.index_put_({nonZeroIndices},
                                   maxValuesNonFinal);
    }
    // compute the expected Q values
    torch::Tensor expectedStateActionValues = (nextStateValues * gamma) + rewards;

    // compute Huber loss
    torch::nn::SmoothL1Loss criterion;
    torch::Tensor loss = criterion(stateActionValues, expectedStateActionValues.unsqueeze(1));

    // optimize the model
    optimizer.zero_grad();

    loss.backward();

    // in-place gradient clipping
    torch::nn::utils::clip_grad_value_(policyNetwork->parameters(), 100);
    optimizer.step();
}

void Agent::train()
{
    // create optimizer for learn()
    torch::optim::AdamW optimizer = torch::optim::AdamW(policyNetwork->parameters(),
                                                        torch::optim::AdamWOptions(learningRate).amsgrad(true));
    if (cleanStart == false)
    {
        loadModel();
    }

    // train for n episodes
    for (int i = 0; i < numEpisodes; i++)
    {
        torch::Tensor state = env->reset();
        while (1)
        {
            // act on the current state
            bool terminated = act(state);

            // learn how action impacted state and optimize
            learn(optimizer);

            // Soft update the target network's weights
            auto targetNetworkStateDictionary = targetNetwork->stateDictionary();
            auto policyNetworkStateDictionary = policyNetwork->stateDictionary();

            for (auto &pair : policyNetworkStateDictionary)
            {
                const std::string &key = pair.first;
                const torch::Tensor &policyTensor = pair.second;
                torch::Tensor &targetTensor = targetNetworkStateDictionary[key];

                targetTensor = (policyTensor * tau) + (targetTensor * (1 - tau));
            }

            targetNetwork->loadStateDictionary(targetNetworkStateDictionary);

            env->render();

            if (terminated)
            {
                // episode finished
                // convert outcomes value to win/loss/draw
                auto to_string = [](int val) -> std::string
                {
                    switch (static_cast<int>(val))
                    {
                    case 1:
                        return "Win";
                    case -1:
                        return "Loss";
                    default:
                        return "Draw";
                    }
                };
                // outcome, # legal moves, # illegal moves,
                if (statsParameters.size() != 0)
                {
                    updateStatsParameters();
                    saveStatistics(statsParameters);
                }

                break;
            }
        }
    }
    saveModel();
}

File: src/agent.py
import math
import random
import io
import os
import re

import torch
import torch.nn as nn
import torch.optim as optim

from dqn import DQN
from memory import ReplayMemory
from ttt_environment import TicTacToeEnvironment
from transition import Transition
from device_manager import DeviceManager


class Agent():
    def __init__(self, n_observations, n_actions, number_episodes=1,
                 memory_size=10000, batch_size=128,
                 epsilon_start=0.9, epsilon_end=0.05, epsilon_decay=1000,
                 tau=0.005, gamma=0.99, learning_rate=0.1,
                 policy_output=None, target_output=None, statistics_output=None,
                 policy_input=None, target_input=None, n_hidden_layers=1, n_neurons=128):
        self.env = TicTacToeEnvironment(
            n_observations, n_actions)
        self.n_observations = n_observations
        self.n_actions = n_actions
        self.number_episodes = number_episodes
        self.memory_size = memory_size
        self.batch_size = batch_size
        self.policy_output = policy_output
        self.target_output = target_output
        self.statistics_output = statistics_output
        self.policy_input = policy_input
        self.target_input = target_input
        self.epsilon_start = epsilon_start
        self.epsilon_end = epsilon_end
        self.epsilon_decay = epsilon_decay
        self.learning_rate = learning_rate
        self.tau = tau
        self.gamma = gamma
        self.steps_taken = 0
        self.explores = 0
        self.exploits = 0
        self.n_hidden_layers = n_hidden_layers
        self.n_neurons = n_neurons
        self.start = 0

        device_manager = DeviceManager.get_instance()
        self.device = device_manager.get_device()

        match = None

        if target_input is not None:
            match = re.search(r"\b\d+\b", target_input)
        
        if match:
            self.start = int(match.group())




    def configure(self):
        self.memory = ReplayMemory(self.memory_size)
        self.policy_network = DQN(
            self.n_observations, self.n_actions, self.n_hidden_layers, self.n_neurons).to(self.device)
        self.target_network = DQN(
            self.n_observations, self.n_actions, self.n_hidden_layers, self.n_neurons).to(self.device)
        self.target_network.load_state_dict(self.policy_network.state_dict())
        self.optimizer = optim.AdamW(
            self.policy_network.parameters(), lr=self.learning_rate, amsgrad=True)

    def load_model(self):
        if self.policy_input is not None:
            with open(self.policy_input, 'rb') as f:
                buffer = io.BytesIO(f.read())
            state_dict = torch.load(buffer)
            self.policy_network.load_state_dict(state_dict)
        if self.target_input is not None:
            with open(self.target_input, 'rb') as f:
                buffer = io.BytesIO(f.read())
            state_dict = torch.load(buffer)
            self.target_network.load_state_dict(state_dict)


    def save_model(self):
        if self.policy_output is not None:
            torch.save(self.policy_network.state_dict(), self.policy_output)
        if self.target_output is not None:
            torch.save(self.target_network.state_dict(), self.target_output)

    def save_statistics(self, i):
        # excluding draw!
        headers = (
            'EpisodeNumber,NumberEpisodes,MemorySize,BatchSize,'
            'EpsilonStart,EpsilonEnd,EpsilonDecay,Tau,Gamma,LearningRate,'
            'WinReward,DrawReward,LossReward,legalMoveReward,IllegalMoveReward,'
            'NumberNeurons', 'NumberHiddenLayers'
            'Explore/Exploit,Outcome,NumberIllegalMoves'
        )

        run_number = i + self.start

        out = (
            f'{run_number},{self.number_episodes},{self.memory.capacity},{self.batch_size},'
            f'{self.epsilon_start},{self.epsilon_end},{self.epsilon_decay},{self.tau},{self.gamma},{self.learning_rate},'
            f'{self.env.win_reward},{self.env.draw_reward},{self.env.loss_reward},'
            f'{self.env.legal_move_reward},{self.env.illegal_move_reward},'
            f'{self.n_neurons},{self.n_hidden_layers},'
            f'{self.explores/(self.exploits + self.explores)},{self.env.outcome},{self.env.illegal_moves}\n'
        )

        if os.path.isfile(self.statistics_output):
            # file exists, append
            with open(self.statistics_output, 'a') as f:
                f.write(out)
        else:
            # file does not exist, create and append
            with open(self.statistics_output, 'w') as f:
                f.write(f'{headers}\n{out}')

    def select_action(self):
        sample = random.random()
        eps_threshold = self.epsilon_end + (self.epsilon_start - self.epsilon_end) * \
            math.exp(-1. * self.steps_taken / self.epsilon_decay)
        self.steps_taken += 1
        if sample > eps_threshold:
            with torch.no_grad():
                self.exploits += 1
                return self.policy_network(self.env.state).max(1)[1].view(1, 1)
        else:
            self.explores += 1
            return torch.tensor([[random.randint(0, self.env.action_space-1)]], device=self.device, dtype=torch.long)

    def act(self):
        action = self.select_action()

        reward = self.env.step(
            action.item())

        return (action, reward)

    def observe_state_change(self, action, reward):
        self.memory.push(self.env.state, action,
                         self.env.next_state, reward)
        
        self.env.state = self.env.next_state

    def optimize_model(self):
        if len(self.memory) < self.batch_size:
            return
        transitions = self.memory.sample(self.batch_size)

        batch = Transition(*zip(*transitions))
        non_final_mask = torch.tensor(tuple(map(lambda s: s is not None,
                                                batch.next_state)), device=self.device, dtype=torch.bool)

        non_final_next_states = torch.cat([s for s in batch.next_state
                                           if s is not None])
        state_batch = torch.cat(batch.state)
        action_batch = torch.cat(batch.action)
        reward_batch = torch.cat(batch.reward)

        #
        state_action_values = self.policy_network(
            state_batch).gather(1, action_batch)
        #
        next_state_values = torch.zeros(self.batch_size, device=self.device, )

        #
        with torch.no_grad():
            next_state_values[non_final_mask] = self.target_network(
                non_final_next_states).max(1)[0]
        # Compute the expected Q values
        expected_state_action_values = (
            next_state_values * self.gamma) + reward_batch

        # Compute Huber loss
        criterion = nn.SmoothL1Loss()
        loss = criterion(state_action_values,
                         expected_state_action_values.unsqueeze(1))


        # Optimize the model
        self.optimizer.zero_grad()
        loss.backward()
        # In-place gradient clipping
        nn.utils.clip_grad_value_(self.policy_network.parameters(), 100)
        self.optimizer.step()


    def update_target_network(self):
        # soft update target network
        target_network_state_dict = self.target_network.state_dict()
        policy_network_state_dict = self.policy_network.state_dict()
        for key in policy_network_state_dict:
            target_network_state_dict[key] = policy_network_state_dict[key] * \
                self.tau + target_network_state_dict[key]*(1-self.tau)
        self.target_network.load_state_dict(target_network_state_dict)

    def train(self):
        if self.policy_input is not None:
            self.load_model()

        for i in range(self.number_episodes):
            self.env.reset()
            self.exploits = 0
            self.explores = 0

            while True:
                # select an action and step in the environment
                action, reward = self.act()
                reward = torch.tensor([reward], device=self.device)

                # self.env.render()

                # observe and remember state change
                self.observe_state_change(action, reward)

                # optimize model from what we learn
                self.optimize_model()

                # soft update target network
                self.update_target_network()

                # if episode has ended
                if self.env.terminated:
                    if self.statistics_output is not None:
                        self.save_statistics(i)
                    break

        if self.policy_output is not None:
            self.save_model()
